test/Npgsql.Tests/Replication/PgOutputReplicationTests.cs

```diff
@@ -0,0 +1,537 @@
+using Npgsql.Replication.Logical.Protocol;
+using NpgsqlTypes;
+using NUnit.Framework;
+
+namespace Npgsql.Tests.Replication
+{
+    public class PgOutputReplicationTests : SafeReplicationTestBase<NpgsqlLogicalReplicationConnection>
+    {
+        [Test]
+        public Task CreateReplicationSlot()
+            => SafeReplicationTest(nameof(CreateReplicationSlot) + "_pgoutput",
+                async (slotName, _) =>
+                {
+                    await using var c = await OpenConnectionAsync();
+                    TestUtil.MinimumPgVersion(c, "10.0", "The Logical Replication Protocol (via pgoutput plugin) was introduced in PostgreSQL 10");
+                    await using var rc = await OpenReplicationConnectionAsync();
+                    var options = await rc.CreateReplicationSlot(slotName);
+
+                    using var cmd =
+                        new NpgsqlCommand($"SELECT * FROM pg_replication_slots WHERE slot_name = '{options.SlotName}'",
+                            c);
+                    await using var reader = await cmd.ExecuteReaderAsync();
+
+                    Assert.That(reader.Read, Is.True);
+                    Assert.That(reader.GetFieldValue<string>(reader.GetOrdinal("slot_type")), Is.EqualTo("logical"));
+                    Assert.That(reader.GetFieldValue<string>(reader.GetOrdinal("plugin")), Is.EqualTo("pgoutput"));
+                    Assert.That(reader.Read, Is.False);
+                });
+
+        [Test(Description = "Tests whether INSERT commands get replicated as Logical Replication Protocol Messages")]
+        public Task StartReplicationReplicatesInsert()
+            => SafeReplicationTest(nameof(StartReplicationReplicatesInsert) + "_pgoutput",
+                async (slotName, tableName, publicationName) =>
+                {
+                    await using var c = await OpenConnectionAsync();
+                    TestUtil.MinimumPgVersion(c, "10.0", "The Logical Replication Protocol (via pgoutput plugin) was introduced in PostgreSQL 10");
+                    var messages = new ConcurrentQueue<LogicalReplicationProtocolMessage>();
+                    await c.ExecuteNonQueryAsync(@$"
+CREATE TABLE {tableName} (id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY, name TEXT NOT NULL);
+CREATE PUBLICATION {publicationName} FOR TABLE {tableName};
+");
+                    var rc = await OpenReplicationConnectionAsync();
+                    var slot = await rc.CreateReplicationSlot(slotName);
+                    await c.ExecuteNonQueryAsync($"INSERT INTO {tableName} (name) VALUES ('val1'), ('val2')");
+
+                    using var streamingCts = new CancellationTokenSource();
+                    var replicationTask = Task.Run(async () =>
+                    {
+                        await foreach (var msg in (await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token)).WithCancellation(streamingCts.Token))
```

Comment:
**IMPORTANT**: it's interesting (and somewhat cumbersome) that we have to pass the cancellation token twice - once to StartReplication, once to WithCancellation. Does it make sense to just perform WithCancellation implicitly, under the hood, before returning the IAsyncEnumerable?

We should research a little bit what the best practice here in .NET - there must be other APIs which return an IAsyncEnumerable, but which are themselves cancellable. Any ideas?

Reply:
> Does it make sense to just perform WithCancellation implicitly, under the hood, before returning the IAsyncEnumerable?

Well, yes and no.
In fact it's actually pretty easy. You need one cancellation token per `await` you want to cancel *Individually*. Because we `await` twice in one line here we also need two cancellation tokens.
It would look more natural if I'd separate out the IAsyncEnumerable<T> here like this:

```csharp
var enumerable = await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token);
await foreach (var msg in enumerable .WithCancellation(streamingCts.Token))
```
We could probably use the cancellation token from the first method in the enumeration loop.
This would be against best practices because you lose the ability to pass different cancellation tokens to the two methods but then again I can't really figure out why that'd be important.

The current (and recommended) syntax even comes with a loaded footgun included.
Both:
```csharp
await foreach (var msg in (await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token)).WithCancellation(streamingCts.Token))
```
and
```csharp
await foreach (var msg in await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token).WithCancellation(streamingCts.Token))
```
are syntactically correct, but only the first one will do what you expect it to do (cancel the stream).

Personally I have no strong feelings either way but please see [this blog post](http://blog.monstuff.com/archives/2019/03/async-enumerables-with-cancellation.html) for how it's meant to be.

Reply:
Thanks for the patient explanation.

I think the loaded footgun is only because we have WithCancellation extension method over `Task<T>` inside Npgsql (which I hate), in normal user code this wouldn't compile.

Thanks for the blog post by Julien Couvreur, that's helpful. The sentence which reminded me of the important thing is:

> the same cancellation token would be used in every enumerator when the collection is enumerated multiple times,

That is, at least in the general case it should be possible to enumerate an IAsyncEnumerable multiple times (just like you can enumerate an IList multiple times), using different cancellation tokens. If we implicitly bake the token passed to StartReplication into the IAsyncEnumerable, that's no longer possible. It's probably a bit weird for people to enumerate the same StartReplication IAsyncEnumerable multiple times, but it does seem better to follow best practices here.

However, check out https://github.com/npgsql/npgsql/pull/3281/files#r515552761 for a suggestion to closely follow the recommended pattern in the blog post, using [EnumeratorCancellation].